u-boot/include/configs/ap152.h

```diff
@@ -0,0 +1,182 @@
+#include <soc/soc_common.h>
+
+#define WLANCAL				0x9fff1000
+#define BOARDCAL			0x9fff0000
+#define ATHEROS_PRODUCT_ID		137
+#define CAL_SECTOR			(CFG_MAX_FLASH_SECT - 1)
+
+/* For Merlin, both PCI, PCI-E interfaces are valid */
+#define ATH_ART_PCICFG_OFFSET		12
+
+#define CONFIG_PCI_CONFIG_DATA_IN_OTP
+
+
+#undef BOARDCAL
+
+
+/*
+ * ==================
+ * GPIO configuration
+ * ==================
+ */
+#if defined(CONFIG_FOR_TPLINK_EAP245_V1)
+	
+	#define CONFIG_QCA_GPIO_MASK_LED_ACT_H	GPIO1 | GPIO7 | GPIO9
+	#define CONFIG_QCA_GPIO_MASK_OUT	GPIO5 | GPIO8 | GPIO10
+	#define CONFIG_QCA_GPIO_MASK_IN		GPIO2
+	#define CONFIG_QCA_GPIO_MASK_OUT_INIT_H GPIO5
+	#define CONFIG_QCA_GPIO_LSUART_TX 17
+	#define CONFIG_QCA_GPIO_LSUART_RX 16
+	#define CONFIG_QCA_GPIO_MDC 8
+	#define CONFIG_QCA_GPIO_MDIO 10
+	#define CONFIG_QCA_MDC_MDIO_AS_GPIO	1
+	#define CONFIG_GPIO_RESET_BTN 2
+#endif
+
+/*
+ * ================
+ * Default bootargs
+ * ================
+ */
+
+#if defined(CONFIG_FOR_TPLINK_EAP245_V1)
+
+	#define CONFIG_BOOTARGS	"console=ttyS0,115200 root=31:02 "\
+				"rootfstype=squashfs init=/sbin/init "\
+				"mtdparts=ath-nor0:128k(u-boot),1024k(kernel),2816k(rootfs),64k(config),64k(art)"
+
+#endif
+
+/*
+ * =============================
+ * Load address and boot command
+ * =============================
+ */
+#if defined(CONFIG_FOR_TPLINK_EAP245_V1)
+	#define CFG_LOAD_ADDR	0x9F040000
+
+#endif
+
+#define CONFIG_BOOTCOMMAND	"bootelf " MK_STR(CFG_LOAD_ADDR)
+
+
+/*
+ * =========================
+ * Environment configuration
+ * =========================
+ */
+#if defined(CONFIG_FOR_TPLINK_EAP245_V1) 
+	#define CFG_ENV_ADDR		0x9F01EC00
+	#define CFG_ENV_SIZE		0x1000
+	#define CFG_ENV_SECT_SIZE	0x10000
+#endif
+
+/*
+ * ===========================
+ * List of available baudrates
+ * ===========================
+ */
+#define CFG_BAUDRATE_TABLE	\
+		{ 600,    1200,   2400,    4800,    9600,    14400, \
+		  19200,  28800,  38400,   56000,   57600,   115200 }
+
+/*
+ * ==================================================
+ * MAC address/es, model and WPS pin offsets in FLASH
+ * ==================================================
+ */
+#if defined(CONFIG_FOR_TPLINK_EAP245_V1)
+	#define OFFSET_MAC_DATA_BLOCK		0x030000
+	#define OFFSET_MAC_DATA_BLOCK_LENGTH	0x010000
+	#define OFFSET_MAC_ADDRESS		0x000008
+	#define OFFSET_ROUTER_MODEL		0x001008
+	#define OFFSET_PIN_NUMBER		0x001108
+#endif
+
+/*
+ * =========================
+ * Custom changes per device
+ * =========================
+ */
+#if defined(CONFIG_FOR_TPLINK_EAP245_V1)
+	#undef CONFIG_CMD_GPIO
+#endif
+
+/*
+ * ===========================
+ * HTTP recovery configuration
+ * ===========================
+ */
+#define WEBFAILSAFE_UPLOAD_KERNEL_ADDRESS	CFG_LOAD_ADDR
+
+//	#define WEBFAILSAFE_UPLOAD_ART_ADDRESS	(CFG_FLASH_BASE + 0x10000)
+
+
+/* Firmware size limit */
+#if defined(CONFIG_FOR_TPLINK_EAP245_V1)
+	#define WEBFAILSAFE_UPLOAD_LIMITED_AREA_IN_BYTES	(192 * 1024)
```

Comment:
decrease 64KB for a case of running tp-link fw. (256 * 1024)

Reply:
offset | size | erasesize | name
-- | -- | -- | --
0 | 20000 | 10000 | u-boot
20000 | 10000 | 10000 | pation-table
30000 | 10000 | 10000 | product-info
40000 | 180000 | 10000 | kernel
001c0000 | 00e00000 | 10000 | rootfs
00fc0000 | 30000 | 10000 | config
00ff0000 | 10000 | 10000 | ART

You want to set it to 256 to preserve product-info ?

Reply:
I think to not overwrite the ART partition...

Reply:
@CodeFetch i had config partition in mind. according to the above partition table it should be set to 512. imagine someone rebuilds tp-link firmware with custom modules that in the end results in  filesize that would occupy whole flash chip - he could have made everything right and the image would fail to boot because there would be no space on flash to extract config to

Reply:
I think there are not many people who install a custom U-Boot, modify and rebuild TP-Link firmware and then get an image so big, that it overwrites the config partition. I don't feel sorry for those lost souls...

Reply:
OK, the meaning of WEBFAILSAFE_UPLOAD_LIMITED_AREA_IN_BYTES is not really clear, it does not start at beginning of flash.

https://github.com/pepe2k/u-boot_mod/blob/7a540a78df6082ca9cc82f72c6a83d34938c2503/u-boot/httpd/httpd.c#L207-L208

Setting it to less than kernel offset is a bug. A comment would be welcome.

if we want to preserve config it must be set to:
0x40000 (kernel offset) + 0x30000 (config size) + 0x10000 (art size) -> 512*1024

Reply:
I understand it like you do: Image must be smaller than FLASH - WEBFAILSAFE_UPLOAD_LIMITED_AREA_IN_BYTES

Thus I assume your calculation is correct.
The reason is that the U-Boot mod supports modified routers and thus you can't define MAX_FIRMWARE_SIZE...

Reply:
hm, i see your change.. but..
the code expects WEBFAILSAFE_UPLOAD_LIMITED_AREA_IN_**BYTES**
are you sure hex offsets sum will work too?

Reply:
As there are brackets around the compiler should take care of it.